activesupport/lib/active_support/notifications/fanout.rb

```diff
@@ -218,6 +218,7 @@ def start(name, id, payload)
           def finish(name, id, payload)
             stack = Thread.current[:_event_stack]
             event = stack.pop
+            event.payload.merge!(payload)
```

Comment:
The `merge!` will not work very well if the payload have nested hashes. I believe we should be using an attribute setter, even if we don't document it. Also it would be good to add a test with nested hashes in the payload.

Reply:
I was looking into these nested `hashes` and I noticed a peculiar behaviour. If we used a nested hash the behaviour worked as expected i.e. keys in the nested hash showed up in the `event.payload` without as having to `deep_merge` them in. I saw then that we actually call `dup` on the payload here: 

https://github.com/rails/rails/blob/ff34f78248e367fdc7d59b42b37610427f7339c8/activesupport/lib/active_support/notifications/instrumenter.rb#L66

 If we want to prevent other subscribers from modifying the `paylaod` for a nested hash then we need to do a `deep_dup` here.  So I've added that change as well.

Reply:
@rafaelfranca and I talked about this irl, but for those following at home. 

 The `deep_dup` method surfaces another problem, in ruby calling a `dup` on a `module` without assigning a module creates a an anonymous module. 

So in the case of something like `ActiveSupport::JSON.dup` we get back ` #<Module:0x00007f9f4531bf00>`,  this causes issue with ActionCable since the ActionCable instrumenter which passes the `coder` in the `event.payload`, the `coder` is the module `ActiveSupport::JSON`.  So after calling `dee_dup` on `{coder: ActiveSupport::JSON} we get back `{coder: #<Module0x..>}, and lose information on the coder. 

There is a bug here that subscribers can modified the nested hashes of other subscribers, but we've chosen to fix this issue outside of this PR since this is a bit tricky. And just keep our current fix as is.

Reply:
Thinking about this a bit more, I think we can remove the `.dup` call entirely. Every `instrument` call pushes a new `event` on to the  thread local stack, so other instrumenters can't modify that event anyway, since they will enqueu their own modified event onto the stack.